### DISCUSSION TITLE
Global options for CommandApp?

### DISCUSSION THREAD

**Discussion by wjrogers** [spectreconsole/spectre.console#657] [Q&A] [answered]
I would like to be able to define global options that apply to all commands configured on an instance of `CommandApp`. For example, a `-c` option that specifies an additional configuration file to load.

I tried defining a base `CommandSettings` subclass that all my commands' settings derive from, but that requires that the global options be passed _after_ the command name. A command line like `app.exe -c config.json cmd` raises an "Unexpected option 'c'" error. Is there another way to accomplish this? If not, will you please consider it for a future release?

**Comment by nils-a** [comment, marked as answer]
@wjrogers you're on the right track. Common/global options are implemented using a common base class. The problem you're facing is that options need to be given after the command. Have you tried invoking your app like `app.exe cmd -c config.json` ?

There's an older issue (#193) discussing the point of the order of commands and options.

There's also another discussion concerning "global options": #643

**Reply by wjrogers** [reply]
Yes, I can make it work by using e.g. `app.exe cmd -c config.json`. However, because the configuration option applies to all commands, ideally I would like a) the option to appear in the top-level help screen and b) the parser to accept this option _before_ the command name. #643 appears to want something similar.

**Reply by wjrogers** [reply]
I finished converting my pilot project from Microsoft.Extensions.CommandLineUtils to Spectre.Console.Cli, and now I think that issue #193 is the real blocker for me. Requiring that options be passed after arguments is too strict a limitation. (It would be a user-facing breaking change in an existing CLI app.)

I really like the features of Spectre.Console in general and the design of Spectre.Console.Cli's branch/command structure in particular, so I'll keep an eye on that issue!
